test/AuctionBidder.test.js

```diff
@@ -0,0 +1,122 @@
+const chai = require("chai")
+
+const expect = chai.expect
+const { to1ePrecision, to1e18 } = require("./helpers/contract-test-helpers")
+
+const { deployMockContract } = require("@ethereum-waffle/mock-contract")
+const AuctionJSON = require("../artifacts/contracts/Auction.sol/Auction.json")
+const CoveragePoolJSON = require("../artifacts/contracts/CoveragePool.sol/CoveragePool.json")
+
+const { BigNumber } = ethers
+
+describe("AuctionBidder", () => {
+  let owner
+
+  let mockAuction
+  let mockCoveragePool
+  let auctionBidder
+
+  before(async () => {
+    owner = await ethers.getSigner(0)
+    bidder = await ethers.getSigner(1)
+
+    mockAuction = await deployMockContract(owner, AuctionJSON.abi)
+    mockCoveragePool = await deployMockContract(owner, CoveragePoolJSON.abi)
+
+    const AuctionBidder = await ethers.getContractFactory("AuctionBidder")
+    auctionBidder = await AuctionBidder.deploy(mockCoveragePool.address)
+  })
+
+  describe("takeOfferWithMin", () => {
+    const divisor = 1000
+    const amountOnOffer = 100 // 100 / 1000 * 100% = 10%
+
+    const amount = to1ePrecision(5, 17)
+    const auctionAmountOutstanding = to1e18(1)
+
+    // 100 * (5 * 10**17) / (1 * 10**18) = 50
+    const portionToSeize = 50 // out of 100
+
+    const minAmountToSeize = to1ePrecision(5, 10)
+
+    beforeEach(async () => {
+      await mockAuction.mock.amountOutstanding.returns(auctionAmountOutstanding)
+      await mockAuction.mock.onOffer.returns(amountOnOffer, divisor) // 10% of the cov pool
+    })
+
+    context(
+      "when an amount to be seized is equal to a minimal desired amount to seize",
+      () => {
+        it("should take an offer", async () => {
+          const amountToSeize = to1ePrecision(5, 10) // 5 * 10**10
+
+          await mockCoveragePool.mock.amountToSeize
+            .withArgs(portionToSeize)
+            .returns(amountToSeize)
+
+          // This is a workouround until Hardhat starts supporting Waffle's
+          // https://ethereum-waffle.readthedocs.io/en/latest/matchers.html#called-on-contract
+          // https://github.com/nomiclabs/hardhat/issues/1135
+          await mockAuction.mock.takeOffer
+            .withArgs(amount)
+            .revertsWithReason("takeOffer should be invoked")
```

Comment:
Wouldn't it be better to use an auction stub that emits an event with `takeOffer` arguments? It looks weird when a case "should take an offer" expects a revert.

Reply:
Yea, that was a hack as hardhat doesn't support `onContractCall` yet. Okay, I've replaced it with a stub here 02e6804